examples/bubble_sort.masm

```diff
@@ -0,0 +1,368 @@
+end
+
+# storeValues() -> void
+#
+# Loads counters `i` and `j` and stores [a,b,...] `operand_stack` elements in mem[i], mem[j]
+proc.storeValues
+    # load `i` counter
+    mem_load.1
+
+    # store value `a` at index mem[i]
+    mem_store
+
+    # load `j` counter
+    mem_load.2
+
+    # store value `b` at index mem[j]
+    mem_store
+end
+
+# sort() -> void
+#
+# Loads values [a,b,...] from `memory` into the `operand_stack` before sorting them
+proc.sort
+    # Load values `a` and `b` into the `operand_stack`
+    exec.loadValues
+
+    # Duplicate and compare them
+    dup.1
+    dup.1
+    lt
+
+    # If a > b then swap and store else store values as is
+    if.true
+        swap
+        exec.swapped
+        exec.storeValues
+    else
+        exec.storeValues
+    end
+
+    # Increment counters to point to the next two elements of the array
+    exec.incrementCounters
+end
+
+# sorted() -> bool
+#
+# Checks if the array has been sorted
+proc.sorted
+    # Array is sorted if counters `i` and `j` are at end of array && `swapped` == 0
+    # We don't need to load both `i` and `j`, using only `j` is sufficient
+    # Load `j` counter and `adv` array length
+    mem_load.2
+    mem_load.4
+
+    # Offset by 5 considering that we have 5 pre-set variables in `memory`
+    # and that memory is linear meaning that array starts at mem[5]
+    add.5
+
+    # Check if they are equal; signifying end of array
+    eq
+
+    # Load `swapped` boolean
+    mem_load.0
+    not
+
+    # Check if both requirements are true
+    and
+    if.true
+        push.0
+    else
+        push.1
+    end
+end
+
+
+# printResult() -> void
+#
+# Adds the array elements to the `operand_stack` before program completion and output
+proc.printResult
+    push.1
+    while.true
+        # Load the `i` counter; at this moment will be equal to `adv`
+        mem_load.1
+
+        # Load from `memory` to `operand_stack` element at location mem[i]
+        mem_load
+
+        # Load `g` counter and increment by 1; signifying that 1 element
+        # has been added to the `operand_stack` from `memory`
+        mem_load.3
+        add.1
+        mem_store.3
+
+        # Decrement counters - going through `memory` in reverse order
+        exec.decrementCounters
+
+        # Load `adv` and `g`
+        mem_load.4
+        mem_load.3
+
+        # Check if `g` == `adv` meaning that we have added all elements to the `operand_stack`
+        eq
+
+        # Exits the loop
+        if.true
+            push.0
+        else
+            push.1
+        end
+    end
+end
+
+begin
+    # Initialise the VM
+    exec.initialise
+    
+    # Loop over the array and sort repeatedly
+    while.true
+        # If looping has been done over the whole array reset counters to start over from the beginning
+        exec.resetCounters
+
+        # Sort elements two-by-two repeatedly
+        exec.sort
+
+        # Check if the array is sorted if so then stop looping
+        exec.sorted
+    end
```

Comment:
note: for production code we also want to have a nice interface. This would be better wrapper around a single `sort` procedure that people can use. Things like `initialise`, `resetCounters`, `sorted` should be treated as implementation details.

Reply:
I believe the code would be both more efficient and more readable if it used the stack more often. Here is one suggestion:

```
#! Sort data
#!
#! Input:
#!   operand_stack: [addr, C, ...]
#!   advice_stack: [n, d0, d1, .., dn, ..]
#! Output: [...]
#!
#! Where:
#!   addr: address were the data will be loaded and sorted in-place
#!   C: a commitment to the data to be sorted
#!   n: the number of elements to be sorted
#!   d0..dn: the data to be sorted
#!
#! Cycles: X
proc.sort
  # save a copy of addr
  dup movdn.6
  # => [C, B, A, addr, C, addr ..]

  # initialize the hasher state
  # TODO: handle padding / input array with odd number of elements
  padw padw padw
  # => [C, B, A, addr, C, addr, ..]

  # load length
  adv_push.1
  # => [len, C, B, A, addr, C, addr, ..]

  # load data 
  # optimization: negate the counter, we are going to loop from `[-n,0)`
  neg
  dup neq.0
  while.true
    movdn.13 # save len (TODO: movup?)
    adv_pipe
    movup.13 add.2 # update len
    dup neq.0
  end

  # clear counter and array end 
  drop movup.13 drop

  # TODO: verify commitment

  # sort
  # 1. copy the addr
  # 2. have a boolean to check if any swaps happened
  # 3. stop when the boolean is false
end
```